Add DRBG random method

richsalz opened the pull request:
Ported from the last FIPS release.
Compiles and passes the self-tests.
Tests not ported.

This is a step along the path. The other RAND-related PR's also need to get approved.  And then I will work on switching this for the default PRNG.

mattcaswell commented:
Not reviewed, but this seems like a great idea.

FdaSilvaYY commented on crypto/rand/drbg_lib.c:
```diff
@@ -0,0 +1,436 @@
+    dctx->app_data = app_data;
+}
+
+size_t RAND_DRBG_get_blocklength(DRBG_CTX *dctx)
```
const DRBG_CTX *dctx

richsalz replied:
fixed.

FdaSilvaYY commented on crypto/rand/drbg_lib.c:
```diff
@@ -0,0 +1,436 @@
+    return dctx->blocklength;
+}
+
+int RAND_DRBG_get_strength(DRBG_CTX *dctx)
```
const DRBG_CTX *dctx

richsalz replied:
fixed.

FdaSilvaYY commented on test/fips_drbgvs.c:
```diff
@@ -0,0 +1,371 @@
+#include "fips_utl.h"
+
+static int dparse_md(char *str)
+	{
```
Old formatting style revival ;) ?

mspncp replied:
I agree, an automatic code reformat could be useful, but preferrable at the end of the review process, not now. A code reformat makes it more difficult to compare Richs changes to the original FIPS code.

richsalz replied:
That test code hasn't been looked at yet.

richsalz replied:
but @mspncp, it will be reformatted and changed before I try to compile it :)  Just like the other code.

mspncp commented on crypto/rand/drbg_lib.c:
```diff
@@ -0,0 +1,436 @@
+    }
+}
+
+static size_t drbg_get_entropy(DRBG_CTX *dctx, unsigned char **pout,
```
The original fips_get_entropy() contained a bug which was reported by Lee D Gibbins <ldgibbons@avaya.com> on rt

	[rt.openssl.org #4055] FIPS Object Module User Guide corrections needed for (*get_entropy)()

(now GitHub Issue #2443) In contrary to the title, it was not an documentation issue but a real implementation bug, as it was confirmed by Steve Marquess in

 + https://mta.openssl.org/pipermail/openssl-dev/2015-October/002950.html
 
(The reply went only to [openssl-dev], not to [rt #4055] so it is missing in GitHub Issue #2443)

Since the bug could not be fixed, because the FIPS code was frozen, Steve Marquess just added a documentation of the bug to section 6.1.1 of the OpenSSL FIPS user guide on 2015-09-30:

	2015-09-30 Section 6.1.1, expanded discussion of the entropy callback (thanks to Lee D Gibbins ldgibbons@avaya.com) 

Now since the code now is being defrosted, this is the opportunity to finally fix this bug.

mspncp commented on crypto/rand/drbg_rand.c:
```diff
@@ -0,0 +1,932 @@
+    return 1;
+}
+
+static int drbg_ctr_generate(DRBG_CTX *dctx,
```
Rich, There is also a bug in drbg_ctr_generate() which I reported privately to Dr. Henson in Sept 2016 and which went unanswered. The report is too long to be reproduced here, so I will send you a copy.

richsalz commented:
None of the test stuff has been reviewed (by me) yet.  It won’t compile.

richsalz commented:
Look forward to your email about the bug.  Hope it has a patch ☺

richsalz commented:
We definitely don't!  Did I leave it in somewhere?

paulidale commented on test/fips_drbgvs.c:
```diff
@@ -0,0 +1,371 @@
+	long entlen, noncelen, perslen, adinlen;
+	int df = 0;
+
+	enum dtype { DRBG_NONE, DRBG_CTR, DRBG_HASH, DRBG_HMAC, DRBG_DUAL_EC }
```
Could we not define dual Ec.

paulidale replied:
I figured out how to comment on a line of code eventually :)

Worth a grep too.

paulidale commented on test/fips_drbgvs.c:
```diff
@@ -0,0 +1,371 @@
+			else if (strstr(buf, "HMAC_DRBG"))
+				drbg_type = DRBG_HMAC;
+			else if (strstr(buf, "Dual_EC_DRBG"))
+				drbg_type = DRBG_DUAL_EC;
```
here too

paulidale commented on crypto/rand/drbg_lib.c:
```diff
@@ -0,0 +1,436 @@
+    default:
+        ret = -2;
+        break;
+    case NID_sha1:
```
Is it worth considering not including the SHA1 based DRBGs?  This and hmacWithSHA1 below.

The known (and foreseeable) weakness in SHA1 is unlikely to impact its use as a DRBG but paranoia is bliss.

richsalz commented:
Good idea, SHA1 removed.

richsalz commented:
Removed DRBG_DUAL_EC

mspncp commented on crypto/rand/drbg_rand.c:
```diff
@@ -0,0 +1,929 @@
+    return 1;
+}
+
+DRBG_CTX *RAND_DRBG_get_default(void)
```
### Default DRBG should be initialized with sensible entropy callbacks

The original FIPS 2.0 DRBG is somewhat incomplete because it implements only the deterministic part and relies on the caller to provide entropy callbacks during FIPS DRBG initialisation. These callbacks were provided by the OpenSSL 1.0.2 crypto library during the FIPS DRBG initialization (see the end of [rand_lib.c](https://github.com/openssl/openssl/blob/OpenSSL_1_0_2-stable/crypto/rand/rand_lib.c))

This code has been removed from master during the FIPS code cleanup in 700b4a4ae7c2a89ca99bfe64baef1eabfa316136. As a consequence, the DRBG_CTX currently returned by RAND_DRBG_get_default() (hence also the RAND_METHOD returned by the RAND_drbg()) is unusable unless the user (i.e., the programmer) provides entropy callbacks.

I think it would be a good idea if do_ossl_drbg_init() would take care of providing sensible defaults. You could take the removed code from 700b4a4ae7c2a89ca99bfe64baef1eabfa316136 as a starting point.

mspncp replied:
Maybe one could link the initialization with the pool concept discussed in #3758: Say, one could distinguish between deterministic methods like rand_drbg_meth, which require being connected to an entropy pool (or whatever it will be called) and non-deterministic methods like rand_os_meth (forwarding to the optimal os rng) which manage their own entropy.
 


By the way: have the naming conventions been loosened since 1.0.2 or shouldn't the function's name be RAND_drbg_method() ?

> RAND_METHOD *RAND_drbg(void)
> {
>     return &rand_drbg_meth;
> }
>

mspncp replied:
Forget my last question: It used to be RAND_SSLeay()  ;-)

richsalz commented:
It’s not ready yet ☺  Tests fail.

richsalz commented:
Ready for review now.
Note that the API *will* be changing.  don't use yet.

levitte commented on util/libcrypto.num:
```diff
@@ -4342,6 +4342,21 @@ OSSL_STORE_INFO_set0_NAME_description   4284	1_1_1	EXIST::FUNCTION:
 OSSL_STORE_LOADER_get0_engine           4287	1_1_1	EXIST::FUNCTION:
-OPENSSL_fork_prepare                    4288	1_1_1	EXIST:UNIX:FUNCTION:
-OPENSSL_fork_parent                     4289	1_1_1	EXIST:UNIX:FUNCTION:
-OPENSSL_fork_child                      4290	1_1_1	EXIST:UNIX:FUNCTION:
```
Not sure why you're moving those...  no biggie, since these symbols are not part of any release yet, but essentially bad form, gives me hickups...

richsalz commented:
I moved them because as I was adding API’s the rebase became tricky.  So I just kept things together.

mspncp commented on crypto/rand/drbg_lib.c:
```diff
@@ -0,0 +1,361 @@
+    }
+}
+
+static size_t get_entropy(DRBG_CTX *dctx, unsigned char **pout,
```
Comparing the the original FIPS code for fips_get_entropy() and your get_entropy() it appears to me that [rt.openssl.org #4055] (Github Issue #2443) hasn't been addressed yet, see my outdated review and https://mta.openssl.org/pipermail/openssl-dev/2015-October/002950.html.

Will this issue be fixed in this PR, or will be there a separate PR?

richsalz commented:
I am most inclined to fix this in the documentation, which will be in another PR.  But I do not have strong feelings about it.  Your view?  Also would love to hear what @snhenson thinks.  Thank you for the reminder!

mspncp commented:
In my opinion, it's clearly a bug. Quoting Lee D Gibbon from his original bug report:

> The "entropy + bl" expression is mixing types, adding bits and bytes together.
> Anyone defining a (*get_entropy)() callback had better ignore the <entropy> parameter. 
> What's more, the callback had better return <min_len> rounded up to a dctx->entropy_blocklen 
> boundary or face failure. 

Steve Marquess agreed that it was a bug in his reply on openssl-dev. The only reason, why the bug was _documented_, not _fixed_, was the fips code freeze. 

IMHO the selftest is useless and can be removed, just as you did in drbg_ctr_generate(). This continous selftest stuff looks weird to me, and I can't find it mentioned in NIST SP80-90A(rev1). Comparing consecutive blocks appears to me a hopelessly naive and simple way of testing whether the entropy source is broken. What if the "entropy source" would just return consecutive integers? The selftest wouldn't notice. (My guess: this test was a special request by the FIPS evaluators?). Also, if a block length of 0 is passed, then the test is skipped anyway.

What is a "block size" of an entropy source in the first place? Is it a hardware constraint of the entropy source, or does it refer to the requirement that entropy(*) should not be added in little portions? 

But, as I said previously: I'm not a crypto expert. You should better rely on @snhenson's opinion.

Regards,
Matthias

 
(*) Even though I'm a mathematician, I don't have any objections to using "entropy" as a synonym for "randomness" in colloquial language. ;-)

mspncp commented:
Not that I object to selftests in general. But the test would have to be  more sophisticated.

richsalz commented:
Selftests are not part of core openssl.  As long as the KAT tests pass I’m good ☺  I’ll try dividing by 8 and see what happens.

mspncp commented:
Divide by 8? Wouldn't it be better to multiply, since the entropy argument (third arg) is in bits?

`
    rv = dctx->get_entropy(dctx, &tout, entropy + 8*bl,
                min_len + bl, max_len + bl);
`

The function signature of the get_entropy() callback is somewhat redundant and error prone, not only due to the mixture of bits and bytes. Why is there a min_len argument if the minimum length of the output must obviously be ceil(min_entropy/8)? A max_len argument makes more sense, since it gives the entropy provider the chance to just add more bytes if the entropy source is not perfect.

Maybe the entire callback signature should be reconsidered?

mspncp commented:
Please have a look at pull request #3899

snhenson commented:
The consecutive block checking was a requirement of FIPS 140-2 which calls it the "Continuous random number generator test". I agree that it's not terribly useful (to put it mildly) but it had to be included for compliance. Since this isn't a requirement here it can be removed.

mspncp commented on crypto/rand/drbg_rand.c:
```diff
@@ -0,0 +1,468 @@
+
+    for ( ; ; ) {
+        inc_128(cctx);
+#if 0
```
I would say yes, it can be removed. But the following occurrances of `lb` and `lb_valid` should then be removed, too:

```
msp@msppc:~/src/openssl/crypto$ rgrep -rw -e 'lb' -e lb_valid rand
rand/drbg_rand.c:304:        if (!dctx->lb_valid) {
rand/drbg_rand.c:305:            AES_encrypt(cctx->V, dctx->lb, &cctx->ks);
rand/drbg_rand.c:306:            dctx->lb_valid = 1;
rand/rand_drbg_lcl.h:110:    unsigned char lb[EVP_MAX_MD_SIZE];
rand/rand_drbg_lcl.h:111:    /* set if lb is valid */
rand/rand_drbg_lcl.h:112:    int lb_valid;

```

richsalz commented:
@mspncp thanks for the reverse engineering and review!  I will address your comments soon (I'm on vacation this week).  Yes, a new PR will swap this into the standard RAND_bytes function :)

richsalz commented:
I removed drbg_check and you already found out that the code was reformatted ☺
About to push additional commits.

paulidale commented on crypto/rand/drbg_lib.c:
```diff
@@ -0,0 +1,336 @@
+        dctx->cleanup_entropy(dctx, out, olen);
+}
+
+
```
Two blanks are twice as good as one?

paulidale commented on crypto/rand/drbg_lib.c:
```diff
@@ -0,0 +1,336 @@
+        ret = dctx->uninstantiate(dctx);
+
+    /* Can't call cleanse because spec requires zeros. */
+    memset(&dctx->ctr, 0, sizeof(dctx->ctr));
```
Is this a FIPS residue?  Why can't we call cleanse?

paulidale commented on crypto/rand/drbg_lib.c:
```diff
@@ -0,0 +1,336 @@
+
+int DRBG_set_ex_data(DRBG_CTX *dctx, int idx, void *arg)
+{
+    return (CRYPTO_set_ex_data(&dctx->ex_data, idx, arg));
```
return (x) -> return x usually.

paulidale commented on crypto/rand/drbg_lib.c:
```diff
@@ -0,0 +1,336 @@
+
+void *DRBG_get_ex_data(const DRBG_CTX *dctx, int idx)
+{
+    return (CRYPTO_get_ex_data(&dctx->ex_data, idx));
```
return (x) -> return x usually.

paulidale commented on crypto/rand/drbg_rand.c:
```diff
@@ -0,0 +1,457 @@
+    DRBG_CTX *dctx = RAND_DRBG_get_default();
+    int ret = 0;
+    unsigned char *adin = NULL;
+    size_t adinlen = 0;
```
These two variables are never used and should be expunged from the code.

paulidale commented on crypto/rand/rand_drbg_lcl.h:
```diff
@@ -0,0 +1,110 @@
+            int entropy, size_t min_len, size_t max_len);
+    /* Indicates we have finished with nonce buffer */
+    void (*cleanup_nonce)(DRBG_CTX *ctx, unsigned char *out, size_t olen);
+
```
Blank line removal?

paulidale commented on test/drbgtest.c:
```diff
@@ -0,0 +1,490 @@
+{
+    if (argc != 1) {
+        TEST_error("Usage: %s", argv[0]);
+        return 0;
```
`return EXIT_SUCCESS;` although `EXIT_FAILURE` might be more appropriate.

richsalz commented:
New commit pushed that addresses all your feedback so far.

mspncp commented:
Oh, you misunderstood my question about the reformatting (it was misleeding).

It was clear to me that you had reformatted the code, but I was wandering why I was getting
so many diffs when running './util/openssl-format-source crypto/rand/drbg_*'.
(You can try it on the current commit ba677f406f.)

But it's nothing important.


On 15.07.2017 14:29, Rich Salz wrote:
> I removed drbg_check and you already found out that the code was reformatted ☺
> About to push additional commits.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/openssl/openssl/pull/3789#issuecomment-315530911>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AK78Tkmu2kXmd4O8tpNBe4BjV5TeWSb-ks5sOLCjgaJpZM4OHNLp>.
>

richsalz commented:
@paulidale I squashed all previous commits and added one more.  It moves function names to private space, removes some unused functions, etc.  Please re-review.

paulidale commented on include/internal/rand.h:
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 1995-2016 The OpenSSL Project Authors. All Rights Reserved.
```
2017 ?

paulidale commented on include/openssl/crypto.h:
```diff
@@ -107,7 +107,8 @@ DEFINE_STACK_OF(void)
 # define CRYPTO_EX_INDEX_UI_METHOD       14
-# define CRYPTO_EX_INDEX__COUNT          15
+# define CRYPTO_EX_INDEX_DRBG            15
+# define CRYPTO_EX_INDEX__COUNT          16
```
What an evil macro name with double underscore.
It's a public API change to fix, so changing it mightn't be possible.

paulidale commented on include/openssl/rand.h:
```diff
@@ -67,7 +67,8 @@ DEPRECATEDIN_1_1_0(int RAND_event(UINT, WPARAM, LPARAM))
 int ERR_load_RAND_strings(void);
 
-# ifdef  __cplusplus
+#ifdef  __cplusplus
```
Shouldn't the space after the # stay?

paulidale replied:
A couple more right above too.

richsalz commented:
The ugly macro name is to discourage public use ☺

I missed renaming one function.  I fixed the space nits you found.  Additional commit pushed; re-review requested.

kroeckx commented:
This doesn't seem to contain any documentation update?

kroeckx commented:
I would really like to see minimal documentation for each function, and this seems to add a lot of new functions without any documentation.

richsalz commented:
That's right, there are no public functions being added.

richsalz commented:
@kroeckx , think of it like the PACKET or WPACKET functions in the SSL library.  These are purely internal functions that will (soon) be called by the existing RAND API.  Users should not call them.  Documentation, since all we have is public manpages, is not appropriate.

kroeckx commented on crypto/rand/drbg_rand.c:
```diff
@@ -0,0 +1,448 @@
+    }
+
+    dctx->max_request = 1 << 16;
+    dctx->reseed_interval = 1 << 24;
```
Is there a reason those values are different than in the SP800-90A?

richsalz replied:
They are what is/was in the OpenSSL FIPS source code.

kroeckx replied:
I don't find this a good answer. At least in the case of max_request it's the difference between bits and bytes.

richsalz replied:
I'm sorry, I dont have a better answer for you; I was just a code-monkey copying our previous FIPS code.  I suggest you ask Steve and/or Andy directly.

mspncp replied:
I think, the value max_request = 2^{16} bytes is correct. On page 49 of [NIST.SP.800-90Ar1.pdf](http://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf)) the bound for max_number_of_bits_per_request = min(B, 2^{19}) bits = 2^{19} bits = 2^{19-3} bytes.

As for the reseed_interval: IMHO the NIST paper describes an abstract mathematical model of an RNG and the OpenSSL DRGB is a concrete implementation of this model. As such its parameters need to remain within the bounds given on page 49. Citing from the bottom of that page:

> Note that the claimed security strength for CTR_DRBG depends on limiting the total number of
> generate requests and the bits provided per generate request according to the table above. 

So in my opinion it is not a violation of the NIST requirements if the DRBG reseeds more often than required. I am using the FIPS DRBG myself and have set the reseed_interval to 1024 requests.

Note: Interestingly, the original FIPS_drbg_set_reseed_interval() in the FIPS Object Module did _not_ contain any bounds check to enforce the NIST bounds. So anyone could call this function from unvalidated code (e.g. from the FIPS capable libcrypto library) to set the reseed interval to an illegal value. And its renamed descendant RAND_DRBG_set_reseed_interval() does not check any bounds either.

kroeckx replied:
I have no problem with making those numbers shorter than what is documented there, but if the numbers are different I have to wonder if there there is a reason why. And "we like it shorter" is a good enough reason for me.

But I think something like RAND_DRBG_set_reseed_interval() should enforce a limit, since we actually have something that says it shouldn't be more than that.

richsalz replied:
that's a good idea, I'll make a new PR.  should it be between 1 and 1<<24 ?

kroeckx replied:
The values from SP800-90A has 2^48 for the AES based ones and 2^32 for the 3 key TDEA. Maybe a parameter in the ctx?

richsalz replied:
Please see https://github.com/openssl/openssl/pull/3970

kroeckx commented on crypto/rand/rand_lcl.h:
```diff
@@ -10,15 +10,93 @@
+# include "include/internal/rand.h"
+
 /* we require 256 bits of randomness */
 # define RANDOMNESS_NEEDED (256 / 8)
```
Where does this fixed value come from?

richsalz replied:
It goes all the way back to openssl 0.9.8.   But 128 or 256 is generally believed to be sufficient if the CSPRNG is good quality.  (Ours wasn't, but this PR is a step on the way to making it so)

kroeckx commented on crypto/rand/drbg_rand.c:
```diff
@@ -0,0 +1,448 @@
+#include "internal/thread_once.h"
+
+/*
+ * Mapping of SP800-90 DRBGs to OpenSSL RAND_METHOD
```
That should probably be "NIST SP 800-90A"

richsalz replied:
Yes, and I will make that change and assume the approval still holds.

kroeckx commented on crypto/rand/drbg_rand.c:
```diff
@@ -0,0 +1,448 @@
+}
+
+/*
+ * Process a complete block using BCC algorithm of SPP 800-90 10.4.3
```
SP, not SPP, and the A is also missing.

kroeckx replied:
And the section is 10.3.3 in revision 1.

richsalz replied:
Wlil update, thanks.

kroeckx commented:
PACKET and WPACKET are documented, as comment in the header files.

kroeckx commented:
It also seems to export a lot of function for something that's internal only.

richsalz commented:
I can add more comments to internal/rand.h in a later PR if you want.  

Yes, there are several new internal functions; many are for the KAT tests, some are for the "deployment model" of the NIST document (instantiate, uninstanstate), and some will be used in the crypto and ssl library.

richsalz commented:
Many comments added, per @kroeckx request.
Sorry to ask for yet another review. :(

richsalz commented:
Whew.  And another step completed, thanks!

kaduk commented on crypto/rand/drbg_rand.c:
```diff
@@ -0,0 +1,449 @@
+        /* Should never happen */
+        n = 16;
+    }
+    for (i = 0; i < 16; i++)
```
Is this loop limit supposed to be 'n' instead of literal 16?  (The previous assignments to 'n' seem to be dead code, otherwise.)

richsalz replied:
Good question.  @snhenson this code comes from the FIPS release.  Is the 16 wrong, or the dead code before the loop?

kaduk replied:
The 16 is wrong; the conceptual CTR_DRBG_Update algorithm assumes the provided input is exactly seedlen bytes long, but our API permits implicit zero padding (since xor with 0s is the identity function).    So we need to only xor as much as it is claimed that we are provided, since I see no API contract in the ctr_XOR/ctr_update usage that implies we will always be passed properly-sized objects.

kaduk commented on crypto/rand/drbg_rand.c:
```diff
@@ -0,0 +1,449 @@
+    *p++ = (inlen >> 8) & 0xff;
+    *p++ = inlen & 0xff;
+
+    /* NB keylen is at most 32 bytes */
```
bytes or bits?

kroeckx replied:
I think we use the key in bytes, and that it's 16, 24 or 32 bytes. But I'm assuming there might be other problems with bits vs bytes problems, as already commented before.

kaduk replied:
Sure.  I just ask because we just assigned to ```*p++``` four times, i.e., consuming 32 bits of storage, having written a length to the buffer being used as AES input.  So this comment would make a lot of sense as saying that we only needed those four octets written.

mspncp commented:
@kaduk you can find the bounds on page 49 of [NIST.SP.800-90Ar1.pdf](http://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf). There the key length is given in bits, here it is in bytes.

@kroeckx The code in this PR is more less the original code from the FIPS object module. And indeed the FIPS module contained a real bug mixing bits and bytes, in the implementation of fips_get_entropy() (This was reported by Lee D Gibbins on rt (See my first review of crypto/rand/drbg_lib.c in this PR for details). But this is the only bits vs. bytes bug I know of.

kaduk commented on include/internal/rand.h:
```diff
@@ -0,0 +1,50 @@
+    void (*cleanup_entropy)(DRBG_CTX *ctx, unsigned char *out, size_t olen),
+    size_t (*get_nonce)(DRBG_CTX *ctx, unsigned char **pout,
+                        int entropy, size_t min_len, size_t max_len),
+    void (*cleanup_nonce)(DRBG_CTX *ctx, unsigned char *out, size_t olen)
```
If I understand correctly, our style seems to be leaning towards typedefs for function types when used as function arguments.  But maybe I understand incorrectly...

richsalz replied:
yes we have been doing that.  note that this PR is closed.  feel free to open and issue or make your own PR

kaduk replied:
Okay, #3971
